### PR TITLE
Compare icds-ucr to citus on dashboard queries

### PR DESCRIPTION
**Pull request by emord**
There's a few todo's here but the general workflow is here. The biggest thing I don't like is that this imports `django.test.override_settings` in tasks.py for the experiment, but the amount of code and hackery it would save makes me think its worth it. Curious how the team feels about it. I couldn't find a good solution to do this with a special router. The other solution I could think of involved making a new manager for citus backend and passing a flag into each query which would be a lot more work than a decorator.

This does not yet handle the exports, which need a separate approach since they're based off of `SqlData`

ignore first commit until https://github.com/dimagi/commcare-hq/pull/24068 is merged

### CODE REVIEW

File: custom/icds_reports/tasks.py [hunk 1]
```diff
@@ -1143,3 +1178,25 @@ def _bust_awc_cache():
...
+    experiment_context = {
+        "function_name": fn.__name__,
+        'args': args,
+        'kwargs': kwargs,
+    }
+    experiment = DashboardQueryExperiment(name="Dashboard Query Experiment", context=experiment_context)
+    with experiment.control() as c:
+        c.record(fn(*args, **kwargs))
+
+    with experiment.candidate() as c:
```
Comment (millerdev) [inline custom/icds_reports/tasks.py]: nit: reusing variable name, single letter variable name 👎 @stickler-ci, where are you?

File: custom/icds_reports/tasks.py [hunk 2]
```diff
@@ -1143,3 +1178,25 @@ def _bust_awc_cache():
...
+        databases = settings.DATABASES
+        databases['icds-ucr'] = deepcopy(databases['citus'])
```
Comment (millerdev) [inline custom/icds_reports/tasks.py]: I think you did not want to overwrite `settings.DATABASES['icds-ucr']` with a copy of the `citus` database (`override_settings` is not even necessary if you do that). You probably wanted this instead
```py
databases = settings.DATABASES.copy()
databases['icds-ucr'] = databases['citus']
```
I think this is a good example of why it's dangerous to be mutating `settings` values in production code.

**Review by millerdev — COMMENTED** [review]
> this imports `django.test.override_settings` in tasks.py

Changing global state with context managers 🤦‍♂️ Do you expect this to be a long-term solution, or is it only for testing a POC and will be removed once the POC is validated?

**Comment by emord** [comment]
> Do you expect this to be a long-term solution, or is it only for testing a POC and will be removed once the POC is validated

Yeah. this whole PR would not last longer than end of June before being removed. (i've been wrong about timelines before, but we're getting pushed hard to get this out so a full scale up can happen ASAP, so there's some benefit for the gov to ensure it happens soon)

Anyways this PR was never likely to get merged in its current state. Mostly presenting my first day's code and figuring out what tradeoffs there are.

Some ideas I've thought of:

1. override databases in context manager 
PROs: very little code, easy to add a decorator to all dbaccessors, can provide all args, kwargs, functions. 
CONs: easy to screw up, modifying global state isn't great
2. [custom manager](https://github.com/dimagi/commcare-hq/pull/24070/commits/268484d5f16ed067f0898ee09f96145120287a93)
PROs: very little code, easy to get all the sql statements
CONs: doesn't give any info on where the query came from (possibly mitigated by passing a traceback), can't provide function arguments, relies on overriding some django internals
3. Pull all SQL queries from django logging and replay them with a task (not possible because django doesn't log queries when `DEBUG = False`
4. log all queries via postgres, create process to scrape the logs from, replay those logs in celery tasks
PROs: Very little complexity in HQ code, doesn't add any performance degredation in python code
CONs: may impact postgres disk performance, would be hard to track it back in the app, would need some other process running on the postgres machines to parse logs
5. Create a new router - I haven't really figured that one out, but I think it would probably involve overriding a setting to force the candidate query to go to postgres. similar pros/cons to #1 without the benefit of being able to provide the function that the query came from

**Comment by millerdev** [comment]
I agree that the first option seems easiest and probably cleanest. We should make sure this technique is not used widely and not for long-term solutions.

**Comment by emord** [comment]
> I agree that the first option seems easiest and probably cleanest. 

@millerdev Curious if you still feel this way after seeing https://github.com/dimagi/commcare-hq/pull/24098. I went with option 2, the custom manager. I think its small enough code, doesn't modify any state, and will hopefully be able to provide enough context to the task to find where its coming from in the code if there are discrepancies

**Comment by millerdev** [comment]
#24098 seems like a reasonable approach. I made a few comments over there.

**Pull request closed without merging**